File: android/app/src/main/java/com/warnabrotha/app/ui/viewmodel/AppViewModel.kt
package com.warnabrotha.app.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.messaging.FirebaseMessaging
import com.warnabrotha.app.data.model.*
import com.warnabrotha.app.data.repository.AppRepository
import com.warnabrotha.app.data.repository.Result
import kotlinx.coroutines.async
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

enum class OTPStep {
    EMAIL_INPUT,
    CODE_INPUT
}

data class AppUiState(
    val isAuthenticated: Boolean = false,
    val isEmailVerified: Boolean = false,
    val showEmailVerification: Boolean = false,
    val otpStep: OTPStep = OTPStep.EMAIL_INPUT,
    val otpEmail: String = "",
    val canResendOTP: Boolean = false,
    val resendCooldownSeconds: Int = 0,
    val parkingLots: List<ParkingLot> = emptyList(),
    val lotStats: Map<Int, ParkingLotWithStats> = emptyMap(),
    val selectedLot: ParkingLotWithStats? = null,
    val selectedLotId: Int? = null,
    val currentSession: ParkingSession? = null,
    val feed: FeedResponse? = null,
    val allFeedSightings: List<FeedSighting> = emptyList(),
    val allFeedsTotalCount: Int = 0,
    val feedFilterLotId: Int? = null, // null means "ALL" is selected
    val totalRegisteredDevices: Int = 0,
    val prediction: PredictionResponse? = null,
    val displayedProbability: Double = 0.0,
    val isLoading: Boolean = false,
    val error: String? = null,
    val successMessage: String? = null,
    // Notification fields
    val pushToken: String? = null,
    val notificationPermissionGranted: Boolean = false,
    val unreadNotificationCount: Int = 0
)

@HiltViewModel
class AppViewModel @Inject constructor(
    private val repository: AppRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(AppUiState())
    val uiState: StateFlow<AppUiState> = _uiState.asStateFlow()

    init {
        if (repository.hasToken() && repository.hasCompletedOnboarding()) {
            _uiState.value = _uiState.value.copy(
                isAuthenticated = true,
                isEmailVerified = true
            )
            silentRefresh()
        } else if (repository.hasToken()) {
            _uiState.value = _uiState.value.copy(isAuthenticated = true)
            silentRefresh()
        }
    }

    private fun silentRefresh() {
        viewModelScope.launch {
            when (val result = repository.register()) {
                is Result.Success -> {
                    if (result.data.emailVerified) {
                        repository.setCompletedOnboarding(true)
                        _uiState.value = _uiState.value.copy(
                            isAuthenticated = true,
                            isEmailVerified = true,
                            showEmailVerification = false
                        )
                        loadInitialData()
                    } else {
                        _uiState.value = _uiState.value.copy(
                            isAuthenticated = true,
                            showEmailVerification = true
                        )
                    }
                }
                is Result.Error -> {
                    // Network error — use existing token optimistically
                    if (repository.hasCompletedOnboarding()) {
                        _uiState.value = _uiState.value.copy(
                            isAuthenticated = true,
                            isEmailVerified = true,
                            showEmailVerification = false
                        )
                        loadInitialData()
                    }
                }
            }
        }
    }

    fun register() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            when (val result = repository.register()) {
                is Result.Success -> {
                    if (result.data.emailVerified) {
                        repository.setCompletedOnboarding(true)
                        _uiState.value = _uiState.value.copy(
                            isAuthenticated = true,
                            isEmailVerified = true,
                            showEmailVerification = false,
                            isLoading = false
                        )
                        loadInitialData()
                    } else {
                        _uiState.value = _uiState.value.copy(
                            isAuthenticated = true,
                            isLoading = false,
                            showEmailVerification = true
                        )
                    }
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = result.message
                    )
                }
            }
        }
    }

    fun sendOTP(email: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null, otpEmail = email)

            when (val result = repository.sendOTP(email)) {
                is Result.Success -> {
                    if (result.data.success) {
                        _uiState.value = _uiState.value.copy(
                            isLoading = false,
                            otpStep = OTPStep.CODE_INPUT
                        )
                        startResendCooldown()
                    } else {
                        _uiState.value = _uiState.value.copy(
                            isLoading = false,
                            error = result.data.message
                        )
                    }
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = result.message
                    )
                }
            }
        }
    }

    fun verifyOTP(otpCode: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            val email = _uiState.value.otpEmail

            when (val result = repository.verifyOTP(email, otpCode)) {
                is Result.Success -> {
                    if (result.data.success && result.data.emailVerified) {
                        repository.setCompletedOnboarding(true)
                        _uiState.value = _uiState.value.copy(
                            isEmailVerified = true,
                            showEmailVerification = false,
                            otpStep = OTPStep.EMAIL_INPUT,
                            isLoading = false
                        )
                        loadInitialData()
                    } else {
                        _uiState.value = _uiState.value.copy(
                            isLoading = false,
                            error = result.data.message
                        )
                    }
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = result.message
                    )
                }
            }
        }
    }

    fun resendOTP() {
        sendOTP(_uiState.value.otpEmail)
    }

    fun changeEmail() {
        _uiState.value = _uiState.value.copy(
            otpStep = OTPStep.EMAIL_INPUT,
            error = null
        )
    }

    private fun startResendCooldown() {
        _uiState.value = _uiState.value.copy(canResendOTP = false, resendCooldownSeconds = 30)
        viewModelScope.launch {
            for (i in 29 downTo 0) {
                delay(1000)
                _uiState.value = _uiState.value.copy(resendCooldownSeconds = i)
            }
            _uiState.value = _uiState.value.copy(canResendOTP = true, resendCooldownSeconds = 0)
        }
    }

    private fun loadInitialData() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            android.util.Log.d("AppViewModel", "Loading initial data...")

            // Load parking lots
            when (val lotsResult = repository.getParkingLots()) {
                is Result.Success -> {
                    val lots = lotsResult.data
                    android.util.Log.d("AppViewModel", "Loaded ${lots.size} parking lots")
                    val selectedLotId = lots.firstOrNull()?.id
                    _uiState.value = _uiState.value.copy(
                        parkingLots = lots,
                        selectedLotId = selectedLotId,
                        feedFilterLotId = null // Default to ALL for feed
                    )
                    selectedLotId?.let { loadLotData(it) }
                    // Load all feeds by default for the feed tab
                    loadAllFeeds()
                    // Load stats for all lots (for map view)
                    loadAllLotStats(lots)
                    // Load global stats (for homepage)
                    loadGlobalStats()
                    // Load unread notification count
                    fetchUnreadNotificationCount()
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to load lots: ${lotsResult.message}")
                    _uiState.value = _uiState.value.copy(error = lotsResult.message)
                }
            }

            // Load current session
            when (val sessionResult = repository.getCurrentSession()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        currentSession = sessionResult.data
                    )
                }
                is Result.Error -> { /* Session might not exist */ }
            }

            _uiState.value = _uiState.value.copy(isLoading = false)
        }
    }

    private fun loadAllLotStats(lots: List<ParkingLot>) {
        viewModelScope.launch {
            val statsMap = mutableMapOf<Int, ParkingLotWithStats>()
            lots.forEach { lot ->
                when (val result = repository.getParkingLot(lot.id)) {
                    is Result.Success -> {
                        statsMap[lot.id] = result.data
                    }
                    is Result.Error -> {
                        android.util.Log.e("AppViewModel", "Failed to load stats for lot ${lot.id}: ${result.message}")
                    }
                }
            }
            _uiState.value = _uiState.value.copy(lotStats = statsMap)
        }
    }

    fun refreshAllLotStats() {
        viewModelScope.launch {
            loadAllLotStats(_uiState.value.parkingLots)
            loadGlobalStats()
        }
    }

    private fun loadGlobalStats() {
        viewModelScope.launch {
            when (val result = repository.getGlobalStats()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        totalRegisteredDevices = result.data.totalRegisteredDevices
                    )
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to load global stats: ${result.message}")
                    // Don't show error to user, just use default value
                }
            }
        }
    }

    fun loadLotData(lotId: Int) {
        viewModelScope.launch {
            // Load lot details
            when (val result = repository.getParkingLot(lotId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(selectedLot = result.data)
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to load lot: ${result.message}")
                }
            }

            // Load prediction
            when (val result = repository.getPrediction(lotId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        prediction = result.data,
                        displayedProbability = result.data.probability
                    )
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to load prediction: ${result.message}")
                }
            }

            // Load feed
            when (val result = repository.getFeed(lotId)) {
                is Result.Success -> {
                    android.util.Log.d("AppViewModel", "Feed loaded: ${result.data.sightings.size} sightings")
                    _uiState.value = _uiState.value.copy(feed = result.data)
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to load feed: ${result.message}")
                }
            }
        }
    }

    fun selectLot(lotId: Int) {
        _uiState.value = _uiState.value.copy(selectedLotId = lotId)
        loadLotData(lotId)
    }

    fun selectFeedFilter(lotId: Int?) {
        _uiState.value = _uiState.value.copy(feedFilterLotId = lotId)
        if (lotId == null) {
            loadAllFeeds()
        } else {
            loadFeedForLot(lotId)
        }
    }

    private fun loadAllFeeds() {
        viewModelScope.launch {
            when (val result = repository.getAllFeeds()) {
                is Result.Success -> {
                    val allSightings = result.data.feeds
                        .flatMap { it.sightings }
                        .sortedBy { it.minutesAgo }
                    _uiState.value = _uiState.value.copy(
                        allFeedSightings = allSightings,
                        allFeedsTotalCount = result.data.totalSightings
                    )
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to load all feeds: ${result.message}")
                }
            }
        }
    }

    private fun loadFeedForLot(lotId: Int) {
        viewModelScope.launch {
            when (val result = repository.getFeed(lotId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(feed = result.data)
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to load feed: ${result.message}")
                }
            }
        }
    }

    fun checkIn() {
        val lotId = _uiState.value.selectedLotId ?: return
        checkInAtLot(lotId)
    }

    fun checkInAtLot(lotId: Int) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            when (val result = repository.checkIn(lotId)) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        currentSession = result.data,
                        isLoading = false,
                        successMessage = "Checked in successfully!"
                    )
                    loadLotData(lotId)
                    refreshAllLotStats()
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = result.message
                    )
                }
            }
        }
    }

    fun checkOut() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            when (val result = repository.checkOut()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        currentSession = null,
                        isLoading = false,
                        successMessage = "Checked out successfully!"
                    )
                    _uiState.value.selectedLotId?.let { loadLotData(it) }
                    refreshAllLotStats()
                }
                is Result.Error -> {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = result.message
                    )
                }
            }
        }
    }

    fun reportSighting(notes: String? = null) {
        val lotId = _uiState.value.selectedLotId ?: return
        reportSightingAtLot(lotId, notes)
    }

    fun reportSightingAtLot(lotId: Int, notes: String? = null) {
        android.util.Log.d("AppViewModel", "reportSightingAtLot called with lotId: $lotId")
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            when (val result = repository.reportSighting(lotId, notes)) {
                is Result.Success -> {
                    android.util.Log.d("AppViewModel", "Report success: ${result.data.usersNotified} users notified")
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        successMessage = "TAPS reported! ${result.data.usersNotified} users notified."
                    )
                    loadLotData(lotId)
                    loadAllFeeds() // Refresh the feed
                    refreshAllLotStats() // Refresh lot stats
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Report error for lotId $lotId: ${result.message}")
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = result.message
                    )
                }
            }
        }
    }

    fun vote(sightingId: Int, voteType: String) {
        viewModelScope.launch {
            // Check the correct data source based on current feed filter
            val filterLotId = _uiState.value.feedFilterLotId
            val currentVote = if (filterLotId == null) {
                _uiState.value.allFeedSightings
                    .find { it.id == sightingId }?.userVote
            } else {
                _uiState.value.feed?.sightings
                    ?.find { it.id == sightingId }?.userVote
            }

            val result = if (currentVote == voteType) {
                // Remove vote (toggle)
                repository.removeVote(sightingId)
            } else {
                // Add or change vote
                repository.vote(sightingId, voteType)
            }

            if (result is Result.Success) {
                // Refresh the correct feed view
                if (filterLotId == null) {
                    loadAllFeeds()
                } else {
                    loadFeedForLot(filterLotId)
                }
            }
        }
    }

    fun refresh() {
        _uiState.value.selectedLotId?.let { loadLotData(it) }
        fetchUnreadNotificationCount()
        viewModelScope.launch {
            when (val sessionResult = repository.getCurrentSession()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(currentSession = sessionResult.data)
                }
                is Result.Error -> { }
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    fun clearSuccessMessage() {
        _uiState.value = _uiState.value.copy(successMessage = null)
    }

    fun onNotificationPermissionResult(granted: Boolean) {
        _uiState.value = _uiState.value.copy(notificationPermissionGranted = granted)
        if (granted) {
            fetchAndSyncPushToken()
        }
    }

    private fun fetchAndSyncPushToken() {
        viewModelScope.launch {
            try {
                val token = FirebaseMessaging.getInstance().token.await()
                updatePushToken(token)
            } catch (e: Exception) {
                android.util.Log.e("AppViewModel", "Failed to get FCM token", e)
            }
        }
    }

    fun updatePushToken(token: String) {
        viewModelScope.launch {
            repository.savePushToken(token)
            _uiState.value = _uiState.value.copy(pushToken = token)
            try {
                repository.updateDevice(pushToken = token, isPushEnabled = true)
            } catch (e: Exception) {
                android.util.Log.e("AppViewModel", "Failed to sync push token with backend", e)
            }
        }
    }

    fun fetchUnreadNotificationCount() {
        viewModelScope.launch {
            when (val result = repository.getUnreadNotifications()) {
                is Result.Success -> {
                    _uiState.value = _uiState.value.copy(
                        unreadNotificationCount = result.data.unreadCount
                    )
                }
                is Result.Error -> {
                    android.util.Log.e("AppViewModel", "Failed to fetch unread count: ${result.message}")
                }
            }
        }
    }
}

File: warnabrotha/warnabrotha/ViewModels/AppViewModel.swift
//
//  AppViewModel.swift
//  warnabrotha
//
//  Main view model managing app state and API interactions.
//

import Foundation
import SwiftUI
import UserNotifications

enum OTPStep {
    case emailInput
    case codeInput
}


@MainActor
class AppViewModel: ObservableObject {
    // MARK: - Published State

    // Auth state
    @Published var isAuthenticated = false
    @Published var isEmailVerified = false
    @Published var showEmailVerification = false

    // OTP state
    @Published var otpStep: OTPStep = .emailInput
    @Published var otpEmail: String = ""
    @Published var canResendOTP = false
    @Published var resendCooldown: Int = 0

    // Parking state
    @Published var parkingLots: [ParkingLot] = []
    @Published var selectedLot: ParkingLotWithStats?
    @Published var currentSession: ParkingSession?
    var isParked: Bool { currentSession != nil }

    // Feed state
    @Published var feed: FeedResponse?
    @Published var allFeeds: AllFeedsResponse?
    @Published var allFeedSightings: [FeedSighting] = []
    @Published var allFeedsLoaded = false

    // Chat state
    @Published var chatMessages: [ChatMessage] = []
    @Published var chatInputText: String = ""
    @Published var chatIsLoading = false
    @Published var chatIsSending = false
    @Published var chatError: String? = nil
    @Published var chatShowError = false

    // Map state
    @Published var lotStats: [Int: ParkingLotWithStats] = [:]
    @Published var globalStats: GlobalStatsResponse? = nil

    // Prediction state
    @Published var prediction: PredictionResponse?
    @Published var displayedProbability: Double = 0
    @Published var isAnimatingProbability = false

    // Notification state
    @Published var notificationPermissionGranted = false
    @Published var unreadNotificationCount = 0
    @Published var showTimeSensitivePrompt = false

    // UI state
    @Published var isLoading = false
    @Published var error: String?
    @Published var showError = false
    @Published var showConfirmation = false
    @Published var confirmationMessage = ""

    // Currently selected lot ID (default to first lot)
    @Published var selectedLotId: Int = 1 {
        didSet {
            UserDefaults.standard.set(selectedLotId, forKey: "selectedParkingLotId")
        }
    }

    private let api = APIClient.shared
    private let keychain = KeychainService.shared
    private var ownMessageIds: Set<Int> = []

    private var resendTimer: Timer?

    var hasCompletedOnboarding: Bool {
        get { UserDefaults.standard.bool(forKey: "hasCompletedOnboarding") }
        set { UserDefaults.standard.set(newValue, forKey: "hasCompletedOnboarding") }
    }

    // MARK: - Initialization

    init() {
        // Keychain persists across app deletions — clear it on a fresh install
        // UserDefaults is wiped on deletion, so this reliably detects reinstalls
        if !UserDefaults.standard.bool(forKey: "hasLaunchedBefore") {
            keychain.clearAll()
            UserDefaults.standard.set(true, forKey: "hasLaunchedBefore")
        }

        if keychain.getToken() != nil && hasCompletedOnboarding {
            isAuthenticated = true
            isEmailVerified = true
            Task {
                await silentRefresh()
            }
        } else if keychain.getToken() != nil {
            isAuthenticated = true
            Task {
                await silentRefresh()
            }
        }

    }

    private func silentRefresh() async {
        do {
            let response = try await api.register()
            if response.emailVerified {
                isEmailVerified = true
                hasCompletedOnboarding = true
                showEmailVerification = false
                await loadInitialData()
                await requestNotificationPermission()
            } else {
                isEmailVerified = false
                showEmailVerification = true
            }
        } catch {
            // Network error — use existing token optimistically
            if hasCompletedOnboarding {
                isEmailVerified = true
                showEmailVerification = false
                await loadInitialData()
            } else {
                // Token exists but email never verified — send back to verification
                showEmailVerification = true
            }
        }
    }

    // MARK: - Authentication

    func register() async {
        isLoading = true
        error = nil

        do {
            let response = try await api.register()
            isAuthenticated = true

            if response.emailVerified {
                isEmailVerified = true
                hasCompletedOnboarding = true
                showEmailVerification = false
                await loadInitialData()
                await requestNotificationPermission()
            } else {
                showEmailVerification = true
            }
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func sendOTP(_ email: String) async {
        isLoading = true
        error = nil
        otpEmail = email

        do {
            let response = try await api.sendOTP(email)
            if response.success {
                otpStep = .codeInput
                startResendCooldown()
            } else {
                self.error = response.message
                showError = true
            }
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func verifyOTP(_ code: String) async {
        isLoading = true
        error = nil

        do {
            let response = try await api.verifyOTP(email: otpEmail, code: code)
            if response.success && response.emailVerified {
                isEmailVerified = true
                hasCompletedOnboarding = true
                showEmailVerification = false
                otpStep = .emailInput
                await loadInitialData()
                await requestNotificationPermission()
            } else {
                self.error = response.message
                showError = true
            }
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func resendOTP() async {
        await sendOTP(otpEmail)
    }

    func changeEmail() {
        otpStep = .emailInput
        error = nil
    }

    private func startResendCooldown() {
        canResendOTP = false
        resendCooldown = 60
        resendTimer?.invalidate()
        resendTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] timer in
            Task { @MainActor in
                guard let self = self else { timer.invalidate(); return }
                self.resendCooldown -= 1
                if self.resendCooldown <= 0 {
                    self.canResendOTP = true
                    timer.invalidate()
                }
            }
        }
    }

    // MARK: - Data Loading

    func loadInitialData() async {
        isLoading = true

        do {
            // Load parking lots
            parkingLots = try await api.getParkingLots()

            // Restore persisted lot selection, or default to first lot
            let savedLotId = UserDefaults.standard.integer(forKey: "selectedParkingLotId")
            if savedLotId != 0, parkingLots.contains(where: { $0.id == savedLotId }) {
                selectedLotId = savedLotId
            } else if let firstLot = parkingLots.first {
                selectedLotId = firstLot.id
            }

            // Check if user is currently parked
            currentSession = try await api.getCurrentSession()

            // Load lot details and prediction
            await loadLotData()

            // Fetch unread notification count
            await fetchUnreadNotificationCount()

            // Load all feeds and global stats
            await loadAllFeeds()
            await loadGlobalStats()
            await refreshAllLotStats()

        } catch let apiError as APIClientError {
            if case .noToken = apiError {
                isAuthenticated = false
            } else {
                self.error = apiError.localizedDescription
            }
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    func loadLotData() async {
        do {
            // Load lot details with stats
            selectedLot = try await api.getParkingLot(id: selectedLotId)

            // Load prediction
            prediction = try await api.getPrediction(lotId: selectedLotId)

            // Animate probability
            animateProbability(to: prediction?.probability ?? 0)

            // Load feed
            feed = try await api.getFeed(lotId: selectedLotId)

        } catch {
            self.error = error.localizedDescription
        }
    }

    func refresh() async {
        await loadLotData()
        await fetchUnreadNotificationCount()
    }

    // MARK: - Notifications

    func requestNotificationPermission() async {
        let granted = await PushNotificationService.shared.requestPermissionAndRegister()
        notificationPermissionGranted = granted
        if granted {
            let settings = await UNUserNotificationCenter.current().notificationSettings()
            if settings.timeSensitiveSetting == .disabled {
                showTimeSensitivePrompt = true
            }
        }
    }

    func fetchUnreadNotificationCount() async {
        do {
            let list = try await api.getUnreadNotifications()
            unreadNotificationCount = list.unreadCount
        } catch {
            // Non-critical — don't surface to UI
            print("Failed to fetch unread notifications: \(error)")
        }
    }

    func markAllNotificationsRead() async {
        guard unreadNotificationCount > 0 else { return }
        do {
            try await api.markAllNotificationsRead()
            unreadNotificationCount = 0
            // Clear the iOS app badge
            try await UNUserNotificationCenter.current().setBadgeCount(0)
        } catch {
            print("Failed to mark all notifications read: \(error)")
        }
    }

    // MARK: - Parking Actions

    func checkIn() async {
        guard !isParked else {
            self.error = "You're already parked!"
            showError = true
            return
        }

        isLoading = true

        do {
            currentSession = try await api.checkIn(lotId: selectedLotId)
            confirmationMessage = "Checked in at \(currentSession?.parkingLotName ?? "parking lot")!"
            showConfirmation = true
            await loadLotData()
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func checkOut() async {
        guard isParked else {
            self.error = "You're not parked!"
            showError = true
            return
        }

        isLoading = true

        do {
            _ = try await api.checkOut()
            let lotName = currentSession?.parkingLotName ?? "parking lot"
            currentSession = nil
            confirmationMessage = "Checked out from \(lotName)!"
            showConfirmation = true
            await loadLotData()
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    // MARK: - Sighting Actions

    func reportSighting(notes: String? = nil) async {
        isLoading = true

        do {
            let parkersAtLot = selectedLot?.activeParkers ?? 0
            _ = try await api.reportSighting(lotId: selectedLotId, notes: notes)
            confirmationMessage = "TAPS reported! \(parkersAtLot) users notified."
            showConfirmation = true
            await loadLotData()
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    // MARK: - Voting

    func vote(sightingId: Int, type: VoteType) async {
        // Save originals for revert on error
        let originalSightings = allFeedSightings
        let currentVote = allFeedSightings.first(where: { $0.id == sightingId })?.userVote

        // Optimistic update — apply immediately before API call
        allFeedSightings = allFeedSightings.map { sighting in
            guard sighting.id == sightingId else { return sighting }
            let newVote: VoteType? = sighting.userVote == type ? nil : type
            var upvotes = sighting.upvotes
            var downvotes = sighting.downvotes
            if sighting.userVote == .upvote { upvotes -= 1 }
            if sighting.userVote == .downvote { downvotes -= 1 }
            if newVote == .upvote { upvotes += 1 }
            if newVote == .downvote { downvotes += 1 }
            return FeedSighting(
                id: sighting.id,
                parkingLotId: sighting.parkingLotId,
                parkingLotName: sighting.parkingLotName,
                parkingLotCode: sighting.parkingLotCode,
                reportedAt: sighting.reportedAt,
                notes: sighting.notes,
                upvotes: upvotes,
                downvotes: downvotes,
                netScore: upvotes - downvotes,
                userVote: newVote,
                minutesAgo: sighting.minutesAgo
            )
        }

        do {
            if currentVote == type {
                try await api.removeVote(sightingId: sightingId)
            } else {
                _ = try await api.vote(sightingId: sightingId, voteType: type)
            }
        } catch {
            // Revert on failure
            allFeedSightings = originalSightings
            self.error = error.localizedDescription
            showError = true
        }
    }

    // MARK: - Map Actions

    func checkInAtLot(_ lotId: Int) async {
        guard !isParked else {
            self.error = "You're already parked!"
            showError = true
            return
        }

        isLoading = true

        do {
            currentSession = try await api.checkIn(lotId: lotId)
            confirmationMessage = "Checked in at \(currentSession?.parkingLotName ?? "parking lot")!"
            showConfirmation = true
            await loadLotData()
            await refreshAllLotStats()
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func reportSightingAtLot(_ lotId: Int, notes: String? = nil) async {
        isLoading = true

        do {
            let parkersAtLot = lotStats[lotId]?.activeParkers ?? selectedLot?.activeParkers ?? 0
            _ = try await api.reportSighting(lotId: lotId, notes: notes)
            confirmationMessage = "TAPS reported! \(parkersAtLot) users notified."
            showConfirmation = true
            await loadLotData()
            await loadAllFeeds()
            await refreshAllLotStats()
        } catch {
            self.error = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    // MARK: - All Feeds

    func loadAllFeeds() async {
        do {
            let response = try await api.getAllFeeds()
            allFeeds = response
            allFeedSightings = response.feeds
                .flatMap { $0.sightings }
                .sorted { $0.minutesAgo < $1.minutesAgo }
            allFeedsLoaded = true
        } catch {
            // Non-critical
            print("Failed to load all feeds: \(error)")
        }
    }

    // MARK: - Global Stats

    func loadGlobalStats() async {
        do {
            globalStats = try await api.getGlobalStats()
        } catch {
            // Non-critical
            print("Failed to load global stats: \(error)")
        }
    }

    // MARK: - Lot Stats

    func refreshAllLotStats() async {
        var statsMap: [Int: ParkingLotWithStats] = [:]
        for lot in parkingLots {
            do {
                let lotWithStats = try await api.getParkingLot(id: lot.id)
                statsMap[lot.id] = lotWithStats
            } catch {
                // Skip this lot
            }
        }
        lotStats = statsMap
    }

    // MARK: - Probability Animation

    private func animateProbability(to target: Double) {
        isAnimatingProbability = true
        let targetPercent = target * 100
        let steps = 30
        let stepDuration = 0.02
        let increment = targetPercent / Double(steps)

        displayedProbability = 0

        Task { @MainActor in
            for i in 1...steps {
                try? await Task.sleep(nanoseconds: UInt64(stepDuration * 1_000_000_000))
                self.displayedProbability = min(increment * Double(i), targetPercent)
            }
            self.isAnimatingProbability = false
        }
    }

    // MARK: - Global Chat

    func loadChatMessages() async {
        chatIsLoading = true
        do {
            let response = try await api.getChatMessages()
            chatMessages = response.messages
        } catch {
            print("Failed to load chat messages: \(error)")
        }
        chatIsLoading = false
    }

    func sendChatMessage() async {
        let content = chatInputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty else { return }

        chatIsSending = true
        let savedInput = chatInputText
        chatInputText = ""

        // Optimistic: show the message immediately while the POST is in flight
        let tempId = -(chatMessages.count + 1)
        let pending = ChatMessage(
            id: tempId,
            content: content,
            sentAt: ISO8601DateFormatter().string(from: Date()),
            minutesAgo: 0,
            isPending: true,
            isOwn: true
        )
        chatMessages.append(pending)

        do {
            let message = try await api.sendChatMessage(content: content)
            ownMessageIds.insert(message.id)
            chatMessages.removeAll { $0.id == tempId }
            if !chatMessages.contains(where: { $0.id == message.id }) {
                var msg = message
                msg.isOwn = true
                chatMessages.append(msg)
            }
        } catch APIClientError.serverError(_, let msg) {
            chatMessages.removeAll { $0.id == tempId }
            chatInputText = savedInput
            chatError = msg
            chatShowError = true
        } catch {
            chatMessages.removeAll { $0.id == tempId }
            chatInputText = savedInput
            chatError = error.localizedDescription
            chatShowError = true
        }

        chatIsSending = false
    }

    // MARK: - Helpers

    var probabilityColor: Color {
        guard let prediction = prediction else { return AppColors.textMuted }
        switch prediction.riskLevel {
        case "HIGH": return AppColors.dangerBright
        case "MEDIUM": return AppColors.warning
        case "LOW": return AppColors.success
        default: return AppColors.textMuted
        }
    }

    var riskLevelText: String {
        guard let prediction = prediction else { return "UNKNOWN" }
        return "\(prediction.riskLevel) RISK"
    }

    var riskMessage: String {
        prediction?.riskMessage ?? "Loading..."
    }

    var riskBars: Int {
        guard let prediction = prediction else { return 2 }
        switch prediction.riskLevel {
        case "HIGH": return 3
        case "MEDIUM": return 2
        case "LOW": return 1
        default: return 2
        }
    }

    func selectLot(_ lotId: Int) {
        selectedLotId = lotId
        Task {
            await loadLotData()
        }
    }
}
